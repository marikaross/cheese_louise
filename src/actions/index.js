export const addCheese = (cheeses) => ({
    type: 'ADD_CHEESE',
    cheeses
})

export const toggleFaveCheese = (cheeseId) => ({
    type: 'TOGGLE_FAVORITE_CHEESE',
    cheeseId
})

export const addFaveCheese = (id) => ({
  type: 'ADD_FAVE_CHEESE',
  id
})

export const deleteCheese = (id) => ({
  type: 'DELETE_CHEESE',
  id
})

export const isLoading = (bool) => ({
  type: 'IS_LOADING',
  isLoading: bool
})

export const hasErrored = (bool) => ({
  type: 'HAS_ERRORED',
  hasErrored: bool
})

export const cheeseFetchDataSuccess = (cheeses) => ({
  type: 'CHEESE_FETCH_DATA_SUCCESS',
  cheeses
})

export const summaryFetchDataSuccess = (summary, id) => ({
  type: 'SUMMARY_FETCH_DATA_SUCCESS',
  id, 
  summary
})
