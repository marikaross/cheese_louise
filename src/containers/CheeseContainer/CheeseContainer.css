.cheeseContainer {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: space-around;
  margin-top: 20px;
}

@media screen and (max-width: 650px) {
  .cheeseContainer {
    display: block;
    margin: 0 auto;
    padding-left: 18%;
  }

}

@media screen and (max-width: 600px) {
  .cheeseContainer {
    padding-left: 12%;
  }

}

@media screen and (max-width: 510px) {
  .cheeseContainer {
    margin: 0 auto;
    padding-left: 0;
  }

}
