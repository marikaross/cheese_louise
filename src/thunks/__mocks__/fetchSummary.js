export const fetchSummary = jest.fn()