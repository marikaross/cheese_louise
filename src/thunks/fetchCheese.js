import { isLoading, hasErrored, cheeseFetchDataSuccess } from '../actions';
import { cheeseCleaner } from '../helper/cheese-cleaner';

export const fetchCheese = (url) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url)
        if(!response.ok) {
          throw Error (response.statusText)
        }
        dispatch(isLoading(false))
        const result = await response.json()
        const cleanCheeses = cheeseCleaner(result)
        dispatch(cheeseFetchDataSuccess(cleanCheeses))
      } catch (error) {
        dispatch(hasErrored(true))
      }
  }

}