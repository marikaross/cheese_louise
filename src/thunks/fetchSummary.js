import {isLoading, hasErrored, summaryFetchDataSuccess } from '../actions';
import { summaryCleaner } from '../helper/cheese-cleaner';

export const fetchSummary = (url, id) => {
  console.log(url)
return async (dispatch) => { 
  try {
    dispatch(isLoading(true))
    const response = await fetch(url)
      if (!response.ok) {
        throw Error (response.statusText)
      }
      dispatch(isLoading(false))
      const result = await response.json()
      console.log(result)
      const cleanSummary = summaryCleaner(result)
      dispatch(summaryFetchDataSuccess(cleanSummary, id))
    } catch (error) {
      dispatch (hasErrored(true))
    }
  }

}