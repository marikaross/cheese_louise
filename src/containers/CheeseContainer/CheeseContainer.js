import React, { Component } from 'react';
import { CheeseCard } from '../../components/CheeseCard';
import { connect } from 'react-redux';
import { addFaveCheese, deleteCheese } from '../../actions';
import { Link } from 'react-router-dom';
import './CheeseContainer.css';
import { fetchSummary } from '../../thunks/fetchSummary';
import PropTypes from 'prop-types'

export class CheeseContainer extends Component{

  toggleFave = (id) => {
    if(this.isDuplicate(id)) {
      this.props.deleteCheese(id)
    } else {
      this.props.addFaveCheese(id)
    }
  }

  isDuplicate = (id) => {
    return this.props.favorites.find(favoriteId => favoriteId === id)
  }

  handleFetch = async (id, urlSnippet) => {
    const url = `https://fr.wikipedia.org/w/api.php?action=opensearch&search=${urlSnippet}&limit=1&format=json`
    await this.props.fetchSummary(url, id)
  }


  cheeseCards = () => {return this.props.cheeses.map(cheese => {
    return (
        <CheeseCard 
          key={cheese.cheeseId}
          name={cheese.name}
          milk={cheese.milk}
          region={cheese.region}
          id={cheese.cheeseId}
          toggleFave={this.toggleFave}
          picture={cheese.picture}
          frenchWiki={cheese.frenchWiki}
          button={cheese.cheeseId}
          handleFetch={this.handleFetch}
        />
      ); 
  })
}

  render() {

    return (
      <div className='cheeseContainer'>
        {this.cheeseCards()}
      </div>
      ) 
  }
}

export const mapStateToProps = (state) => ({
  cheeses: state.cheeses,
  favorites: state.favorites
})

export const mapDispatchToProps = (dispatch) => ({
  deleteCheese: (id) => dispatch(deleteCheese(id)),
  addFaveCheese: (id) => dispatch(addFaveCheese(id)),
  fetchSummary: (summary, id) => dispatch(fetchSummary(summary, id))
})

CheeseContainer.propTypes = {
  deleteCheese: PropTypes.func,
  addFaveCheese: PropTypes.func,
  fetchSummary: PropTypes.func,
  cheeses: PropTypes.array,
  favorites: PropTypes.array
};

export default connect(mapStateToProps, mapDispatchToProps)(CheeseContainer);