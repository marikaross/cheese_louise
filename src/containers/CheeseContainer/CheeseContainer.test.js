import { CheeseContainer, mapStateToProps, mapDispatchToProps } from './CheeseContainer';
import React from 'react';
import { shallow } from 'enzyme';
import * as action from '../../actions';
import { fetchSummary } from '../../thunks/fetchSummary'

describe('CheeseContainer', () => {
  let wrapper
  let mockCheeses
  let mockFavorites
  let mockAddFaveCheese
  let mockDeleteCheese
  let mockFetchSummary

  beforeEach(() => {
    mockCheeses = [{name: 'nacho', id: 6}, {name: 'swiss', id: 5}]
    mockFavorites = [{name: 'nacho', id: 6}]
    mockAddFaveCheese = jest.fn()
    mockDeleteCheese = jest.fn()
    mockFetchSummary = jest.fn()

    wrapper = shallow(
      <CheeseContainer 
        cheeses={mockCheeses}
        favorites={mockFavorites}
        addFaveCheese={mockAddFaveCheese}
        deleteCheese={mockDeleteCheese}
      />)
  })

  it('should match the snapShot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('mapStateToProps', () => {
    it('should return a cheeses array, a favorites array', () => {
      const mockState = {
        cheeses: mockCheeses,
        favorites: mockFavorites,
        addFaveCheese: mockAddFaveCheese
      }

      const expectedProps = {
        cheeses: mockCheeses,
        favorites: mockFavorites
      }

      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expectedProps)
    })
  }) 

  describe('mapDispatchToProps', () => {
    it('calls dispatch with deleteCheese', () => {
      const mockState = {
        cheeses: mockCheeses,
        favorites: mockFavorites,
        addFaveCheese: mockAddFaveCheese,
        deleteCheese: mockDeleteCheese
      }

      const mockDispatch = jest.fn()
      const mappedProps = mapDispatchToProps(mockDispatch)
      const actionToDispatch = action.deleteCheese({id: 5})
      mappedProps.deleteCheese({id: 5})
      expect(mockDispatch).toBeCalledWith(actionToDispatch)
    })

    it('calls dispatch with addFaveCheese', () => {
       const mockState = {
        cheeses: mockCheeses,
        favorites: mockFavorites,
        addFaveCheese: mockAddFaveCheese,
        deleteCheese: mockDeleteCheese
      }

      const mockDispatch = jest.fn()
      const mappedProps = mapDispatchToProps(mockDispatch)
      const actionToDispatch = action.addFaveCheese({id: 8})
      mappedProps.addFaveCheese({id: 8})
      expect(mockDispatch).toBeCalledWith(actionToDispatch)
    })

    it.skip('calls dispatch with addSummary', () => {
       const mockState = {
        cheeses: mockCheeses,
        favorites: mockFavorites,
        addFaveCheese: mockAddFaveCheese,
        deleteCheese: mockDeleteCheese,
        addSummary: mockFetchSummary
      }
      const mockDispatch = jest.fn()
      const mappedProps = mapDispatchToProps(mockDispatch)
      const actionToDispatch = fetchSummary('cheese stuff', 8)
      mappedProps.fetchSummary('cheese stuff', 8)
      expect(mockDispatch).toBeCalledWith(actionToDispatch)
    })
  })
})

