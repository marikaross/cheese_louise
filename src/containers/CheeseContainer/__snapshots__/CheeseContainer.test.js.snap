// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheeseContainer should match the snapShot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    addFaveCheese={[Function]}
    cheeses={
        Array [
            Object {
              "id": 6,
              "name": "nacho",
            },
            Object {
              "id": 5,
              "name": "swiss",
            },
          ]
    }
    deleteCheese={[Function]}
    favorites={
        Array [
            Object {
              "id": 6,
              "name": "nacho",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Unknown
          button={undefined}
          frenchWiki={undefined}
          handleFetch={[Function]}
          id={undefined}
          milk={undefined}
          name="nacho"
          picture={undefined}
          region={undefined}
          toggleFave={[Function]}
/>,
        <Unknown
          button={undefined}
          frenchWiki={undefined}
          handleFetch={[Function]}
          id={undefined}
          milk={undefined}
          name="swiss"
          picture={undefined}
          region={undefined}
          toggleFave={[Function]}
/>,
      ],
      "className": "cheeseContainer",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "button": undefined,
          "frenchWiki": undefined,
          "handleFetch": [Function],
          "id": undefined,
          "milk": undefined,
          "name": "nacho",
          "picture": undefined,
          "region": undefined,
          "toggleFave": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "button": undefined,
          "frenchWiki": undefined,
          "handleFetch": [Function],
          "id": undefined,
          "milk": undefined,
          "name": "swiss",
          "picture": undefined,
          "region": undefined,
          "toggleFave": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Unknown
            button={undefined}
            frenchWiki={undefined}
            handleFetch={[Function]}
            id={undefined}
            milk={undefined}
            name="nacho"
            picture={undefined}
            region={undefined}
            toggleFave={[Function]}
/>,
          <Unknown
            button={undefined}
            frenchWiki={undefined}
            handleFetch={[Function]}
            id={undefined}
            milk={undefined}
            name="swiss"
            picture={undefined}
            region={undefined}
            toggleFave={[Function]}
/>,
        ],
        "className": "cheeseContainer",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "button": undefined,
            "frenchWiki": undefined,
            "handleFetch": [Function],
            "id": undefined,
            "milk": undefined,
            "name": "nacho",
            "picture": undefined,
            "region": undefined,
            "toggleFave": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "button": undefined,
            "frenchWiki": undefined,
            "handleFetch": [Function],
            "id": undefined,
            "milk": undefined,
            "name": "swiss",
            "picture": undefined,
            "region": undefined,
            "toggleFave": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
